### PR TITLE
LYN-7547 | Focus Mode - It is possible to create a child entity of a closed container

### PR DESCRIPTION
**Pull request by AMZN-daimini**
Prefab Focus Mode is meant to disable editing a prefab when it's closed. Right now it's possible to create an entity under a closed prefab container, which does not follow the pattern.

After this fix, it's now no longer possible to create entities or drag existing entities into closed prefabs. Similarly, it's not possible to drag assets from the asset browser onto closed prefabs.

NoNewEntityOnClosedContainer

### CODE REVIEW

**Review by AMZN-Igarri — APPROVED** [review]
LGTM

File: Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp [hunk 1]
```diff
@@ -658,18 +662,19 @@ void SandboxIntegrationManager::PopulateEditorGlobalContextMenu(QMenu* menu, con
...
-            });
+        auto containerEntityInterface = AZ::Interface<AzToolsFramework::ContainerEntityInterface>::Get();
+        if (!prefabSystemEnabled || (containerEntityInterface && containerEntityInterface->IsContainerOpen(selected.front())))
+        {
+            action = menu->addAction(QObject::tr("Create entity"));
+            QObject::connect(
+                action, &QAction::triggered, action,
+                [selected]
+                {
+                    EBUS_EVENT(AzToolsFramework::EditorRequests::Bus, CreateNewEntityAsChild, selected.front());
```
Comment (hultonha) [inline Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp:673]: nit: While we're here could we update this EBus call to not use the old macro syntax

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp [hunk 2]
```diff
@@ -764,10 +765,18 @@ namespace AzToolsFramework
         return canHandleData;
     }
 
-    bool EntityOutlinerListModel::CanDropMimeDataAssets(const QMimeData* data, Qt::DropAction /*action*/, int /*row*/, int /*column*/, const QModelIndex& /*parent*/) const
+    bool EntityOutlinerListModel::CanDropMimeDataAssets(const QMimeData* data, Qt::DropAction /*action*/, int /*row*/, int /*column*/, const QModelIndex& parent) const
     {
         using namespace AzToolsFramework;
```
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp:770]: question: Could we remove this `using namespace` as we're here and instead use a using declaration for the types we care about

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp [hunk 3]
```diff
@@ -764,10 +765,18 @@ namespace AzToolsFramework
         return canHandleData;
     }
 
-    bool EntityOutlinerListModel::CanDropMimeDataAssets(const QMimeData* data, Qt::DropAction /*action*/, int /*row*/, int /*column*/, const QModelIndex& /*parent*/) const
+    bool EntityOutlinerListModel::CanDropMimeDataAssets(const QMimeData* data, Qt::DropAction /*action*/, int /*row*/, int /*column*/, const QModelIndex& parent) const
```
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp:768]: suggestion: Could update these to `[[maybe_unused]]` while we're changing this line (also could break at opening paren as this is already quite a long line)

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp [hunk 4]
```diff
@@ -764,10 +765,18 @@ namespace AzToolsFramework
...
     {
         using namespace AzToolsFramework;
-        
+
+        // Disable dropping assets on closed container entities.
+        AZ::EntityId parentId = GetEntityFromIndex(parent);
+        if (auto containerEntityInterface = AZ::Interface<ContainerEntityInterface>::Get();
```
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp:774]: suggestion: As `auto containerEntityInterface = AZ::Interface<ContainerEntityInterface>::Get();` is called twice we could pull this out and use in both conditionals here and below?

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp [hunk 5]
```diff
@@ -788,6 +797,13 @@ namespace AzToolsFramework
...
 
+        // If the parent entity is a closed container, bail.
+        if (auto containerEntityInterface = AZ::Interface<ContainerEntityInterface>::Get();
+            !containerEntityInterface->IsContainerOpen(assignParentId))
+        {
+            return false;
+        }
+
         // Source Files
         if (sourceFiles.size() > 0)
```
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp:808]: nit: Could update this to `!sourceFiles.empty()` while we're here 😛

**Review by hultonha — APPROVED** [review]
Change looks good, I just have a few minor suggestions. One other question is do we have any tests for this? Is this something we could investigate in a follow-up PR? Might be something to consider to verify this functionality.

File: Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp (hunk 1 above)
Comment (AMZN-daimini) [reply Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp:673]: Done!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp (hunk 3 above)
Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp:768]: Fixed!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp (hunk 2 above)
Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp:770]: Fixed!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp (hunk 5 above)
Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp:808]: Done!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp (hunk 4 above)
Comment (AMZN-daimini) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp:774]: The calls are in separate functions, so I'd have to move it to a member. I was considering doing it, but the calls are very sporadic so I didn't want to complicate the class yet. If I need to add more calls for other bugfixes, I'll consider moving the Get so it's only done once :)

**Review by cgalvan — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp (hunk 4 above)
Comment (hultonha) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/Outliner/EntityOutlinerListModel.cpp:777]: Sounds good 👍

**Pull request merged**